examples/gno.land/p/demo/membstore/membstore.gno

```diff
@@ -0,0 +1,209 @@
+	ErrInvalidAddressUpdate = errors.New("invalid address update")
+	ErrNotGovDAO            = errors.New("caller not correct govdao instance")
+)
+
+// maxRequestMembers is the maximum number of
+// paginated members that can be requested
+const maxRequestMembers = 50
+
+type Option func(*MembStore)
+
+// WithInitialMembers initializes the member store
+// with an initial member list
+func WithInitialMembers(members []Member) Option {
+	return func(store *MembStore) {
+		for _, m := range members {
+			memberAddr := m.Address.String()
+
+			// Check if the member already exists
+			if store.members.Has(memberAddr) {
+				panic(ufmt.Errorf("%s, %s", memberAddr, ErrAlreadyMember))
+			}
+
+			store.members.Set(memberAddr, m)
+			store.totalVotingPower += m.VotingPower
+		}
+	}
+}
+
+// WithDAOPkgPath initializes the member store
+// with a dao package path guard
+func WithDAOPkgPath(daoPkgPath string) Option {
+	return func(store *MembStore) {
+		store.daoPkgPath = daoPkgPath
+	}
+}
+
+// MembStore implements the dao.MembStore abstraction
+type MembStore struct {
+	daoPkgPath       string    // active dao pkg path, if any
+	members          *avl.Tree // std.Address -> Member
+	totalVotingPower uint64    // cached value for quick lookups
+}
+
+// NewMembStore creates a new member store
+func NewMembStore(opts ...Option) *MembStore {
+	m := &MembStore{
+		members:          avl.NewTree(), // empty set
+		daoPkgPath:       "",            // no dao guard
+		totalVotingPower: 0,
+	}
+
+	// Apply the options
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	return m
+}
+
+// AddMember adds member to the member store `m`.
+// It fails if the caller is not GovDAO or
+// if the member is already present
+func (m *MembStore) AddMember(member Member) error {
+	if !m.isCallerDAORealm() {
+		return ErrNotGovDAO
+	}
+
+	// Check if the member exists
+	if m.IsMember(member.Address) {
+		return ErrAlreadyMember
+	}
+
+	// Add the member
+	m.members.Set(member.Address.String(), member)
+
+	// Update the total voting power
+	m.totalVotingPower += member.VotingPower
+
+	return nil
+}
+
+// UpdateMember updates the member with the given address.
+// Updating fails if the caller is not GovDAO.
+func (m *MembStore) UpdateMember(address std.Address, member Member) error {
+	if !m.isCallerDAORealm() {
+		return ErrNotGovDAO
+	}
+
+	// Get the member
+	oldMember, err := m.Member(address)
+	if err != nil {
+		return err
+	}
+
+	// Check if this is a removal request
+	if member.VotingPower == 0 {
+		m.members.Remove(address.String())
+
+		// Update the total voting power
+		m.totalVotingPower -= oldMember.VotingPower
+
+		return nil
+	}
+
+	// Check that the member wouldn't be
+	// overwriting an existing one
+	isAddressUpdate := address != member.Address
+	if isAddressUpdate && m.IsMember(member.Address) {
+		return ErrInvalidAddressUpdate
+	}
+
+	// Remove the old member info
+	// in case the address changed
+	if address != member.Address {
+		m.members.Remove(address.String())
```

Comment:
if it's meant to be changing address of a member, it is better to have another func rather than `Update`. for me, update is more for update voting power, desc, etc. while changing address is more about transfer ownership. WDYT.